### ISSUE TITLE
Unhealthy tasks should be replaced immediately without waiting for the unhealthy task to die

### ISSUE THREAD

**Issue by felixb** [mesosphere/marathon#2104]
Marathon is does _not_ replace unhealthy tasks, if the task is running on an unresponsive mesos slave.

I think, this is because, marathon tries to kill the task before launching a new one.
But because the slave/executor is not responding, it is unable to do so.

This is kind of dangerous in combination with servicerouter in cases, where the slave got removed completely. See #2083.

Steps to reproduce:
1. Launch a task with health check on a mesos slave.
2. Kill that slave.
3. See marathon switching the state to unhealthy but doing nothing about it.

If you set a mesos `--slave_removeal_limit` this even gets worse, because the deactivated slave keeps hanging around for hours breaking marathon and servicerouter.

**Comment by pierlo-upitup** [comment]
@felixb Thanks for reporting this. We are going to take a look at this asap.

**Comment by aquamatthias** [comment]
Hey @felixb that indeed is a problem:

- The HealthCheckManager periodically performs health checks. If it detects consecutive failures it will kill the task via mesos scheduler API.
- The TaskTracker is an independent separate entity and watches the instance count. If the number of desired instances drops, it will restart another one.

Since  the task kill (as requested from the HealthCheckManager) is not executed, the number of instances does not drop and the TaskTracker will not start another task.

I see two points we should work on:
- Ask the mesos guys if it is possible to improve the mechanism of killing a task (even if the slave is not responsive). 
- Improve the servicerouter to only accept healthy tasks

@tillt: do you have any suggestion for the mesos part?
@lloesche: does the servicerouter includes only healthy tasks?

**Comment by felixb** [comment]
Well, I think removing the tasks in servicerouter does not really help a lot. At least we are running health checks in haproxy anyway. The tasks won't get any requests.

The problem is, that no new (healthy) task is spawned.
What do you think about automatically over-provision apps in such situations?

**Comment by bernd-mesos** [comment]
Do you get any TASK_LOST coming from the master notifications in the log?

**Comment by felixb** [comment]
I don't think so.

I removed the mesos flag for limiting the slave removal to mitigate this issue.
Mesos drops the slaves pretty fast now, so marathon receives the TASK_LOST and drops the task.

**Comment by bernd-mesos** [comment]
As long as the master believes there is a chance that a slave can reappear, it will not send TASK_LOST. For that period of time you should experience exactly what is described in this ticket.

**Comment by felixb** [comment]
That's true for mesos's view. But marathon is experiencing failed health checks. It should be able to spawn new task as a replacement, shouldn't it?

**Comment by bernd-mesos** [comment]
Marathon does not progress to spawning a new task before having killed the old one. And killing it must fail, since there is no slave. 

Proposal: Marathon should not try to kill these tasks, but move on immediately.

**Comment by aquamatthias** [comment]
@felixb tried to reproduce this with default mesos master/slave settings:
- start master
- start slave
- start marathon
- start app with health checks, that fail
- wait for all tasks to start
- kill the slave (kill -9)
- after ~75 seconds the master declares all tasks as lost.
- marathon removes those tasks and try to start them on another slave

Everything works as expected. 
Do we speak about the 75 seconds delay? (From the description above it sounds like hours...)

**Comment by felixb** [comment]
It's not about the 75s. mesos `--slave_removal_rate_limit` and `--slave_reregister_timeout` can lead to very long timeouts for slave removal.

**Comment by bernd-mesos** [comment]
If it's 75 seconds, then everything works as expected and no change to Marathon is necessary.

**Comment by aquamatthias** [comment]
@felixb ok just to make sure I understand your suggestion correctly: if a health check marks a task as unhealthy, a new task should be started without waiting until the unhealthy one is killed. That holds true even if the instance count is violated. Correct?

**Comment by felixb** [comment]
Yes. That's what I expect for any kind of web app or similar things.

**Comment by aquamatthias** [comment]
@felixb changed the title to address that requirement and made it an enhancement.

**Comment by drcrallen** [comment]
This happened in my staging cluster as I'm evaluating Mesos.

Right now I'm sitting with 3 tasks for a particular application that are listed as "Unhealthy". Any call looking at the tasks shows `TASK_LOST`. But no matter what I do, I *cannot* get rid of the tasks. This means the application is completely unusable. I cannot scale ore deploy it.

**Comment by lloesche** [comment]
@drcrallen Even after 75 seconds? Does the app have any constraints that would prohibit Marathon from finding a matching resource offer? Like for instance do you have a constraint hostname:UNIQUE. If so you might be running into `--task_lost_expunge_gc` (24h by default).
If that's the case and your cluster is relatively small (like num agents = num tasks) it'd be better to use a GROUP_BY constraint for that app. Then to work around it right now you could recreate the app under a new name (just edit the config of the existing app, go to json mode and change the id) and destroy or scale the old one to 0 (the deployment will run until the task_lost_expunge_gc timeout is reached so don't worry if it's not immediately doing something). That should recover your app right away.

**Comment by viveksatasiya** [comment]
This simmilar thing happened in my testing environment. I have also reported issue. here is the link [https://github.com/mesosphere/marathon/issues/4288](url)

Is there any other way to mitigate this issue??

Thanks

**Comment by bpiper** [comment]
My experience with this in a fairly simple scenario (app is running with single instance and then I terminate the slave node it's running on) is that the task will sit in TASK_KILLED state until such time as the dead node disappears from the list of nodes (i.e. Mesos gives up on it, which seems to take anywhere from a few minutes to 15+ minutes in a vanilla DC/OS cluster), which is the point at which it changes to TASK_LOST and Marathon only then starts a new instance.  

Since Marathon has apparently given up on the task (which it does in a reasonably short time-frame) when it goes to TASK_KILLED, it really should be starting a new one immediately (this possibly should be configurable, but seems like sensible default behaviour).  What's confusing about TASK_KILLED though is that if, rather than terminating my slave node, I just block all network traffic to it for a time (long enough to go to TASK_KILLED state, and for the slave node to show as unhealthy), and then unblock, the same task will go back to being healthy and all is forgiven.

It seems wrong to imply that the task is killed if there was no way of contacting the slave to actually kill the task, and if we're prepared to accept its resurrection.  I think in many (maybe even most) cases, it would be desirable to start a new task, and blacklist the old one in such a way that it gets properly killed if it ever comes back to life, but otherwise just ignored.

**Comment by meichstedt** [comment]
The related functionality has changed in Marathon 1.4:

- Marathon will not perform HealthChecks for tasks that are unreachable (hence healthChecks won't fail)
- When a task becomes unreachable (due to the agent being disconnected/removed), Marathon will consider the task inactive after `unreachableStrategy.inactiveAfter` (defaulting to 5 minutes for normal tasks). That means, it will try to launch a new task after that timeout, and eventually expunge information about the unreachable task when it's been unreachable for longer than `unreachableStrategy.expungeAfter` (by default 10 minutes).

[UnreachableStrategy](https://github.com/mesosphere/marathon/blob/8016730b9ea7379399229f4b464a675281e33f0f/src/main/scala/mesosphere/marathon/state/UnreachableStrategy.scala) can be defined for each app/pod in Marathon 1.4.

**Comment by bpiper** [comment]
That sounds like just what the doctor ordered... thanks for the heads up @meichstedt :+1:

**Comment by heww** [comment]
@meichstedt but today I got another result. 

At first I have a app with one instance run in Marathon 1.3.10.
Today, I upgrade Marathon to 1.4.1.
At 17:00 (9:00 utc), the instance health check become to unknow, but after 30 minutes, instance still unknow, no new task launch. 

My website become unreachable through marathon-lb. After that i scale instances to 2.


![image](https://cloud.githubusercontent.com/assets/5634205/23165727/80e3cc98-f878-11e6-8410-8621f60866d2.png)

app's config of unreachableStrategy is below
```
    "unreachableStrategy": {
      "inactiveAfterSeconds": 300,
      "expungeAfterSeconds": 600
    },
    "killSelection": "YOUNGEST_FIRST",
```

**Comment by jdef** [comment]
Is your app using any resident resources?

**Comment by heww** [comment]
what is resident resources mean? if it is  local persistent volumes document in 
https://mesosphere.github.io/marathon/docs/persistent-volumes.html ,  i think we don't use it.

my app's volumes is like this, and /mnt/container-volumes mount a nfs at every mesos agent.
```json
"volumes": [
	{
		"containerPath": "/app/media",
		"hostPath": "/mnt/container-volumes/foo/app/media",
		"mode": "RW"
	}
]
```

**Comment by meichstedt** [comment]
@heww thanks for reporting this. It's actually another issue, I filed a separate bug for this: #5235

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-3579. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
